File: tools/lib/confluence.py
from __future__ import annotations

import base64
import html
import json
from typing import Any

import requests


def _basic_auth(email: str, token: str) -> str:
    raw = f"{email}:{token}".encode("utf-8")
    return "Basic " + base64.b64encode(raw).decode("ascii")


def handshake(base_url: str, email: str, api_token: str) -> dict[str, Any]:
    base = base_url.rstrip("/")
    url = f"{base}/rest/api/user/current"
    r = requests.get(
        url,
        headers={"Accept": "application/json", "Authorization": _basic_auth(email, api_token)},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def markdown_to_storage_html(md: str) -> str:
    """Minimal Markdown → Confluence storage: paragraphs and **bold**."""
    lines: list[str] = []
    for block in md.replace("\r\n", "\n").split("\n\n"):
        block = block.strip()
        if not block:
            continue
        out = []
        for line in block.split("\n"):
            parts = line.split("**")
            if len(parts) % 2 == 1:
                esc = [html.escape(p) for p in parts]
                line_html = "".join(
                    esc[i] if i % 2 == 0 else f"<strong>{esc[i]}</strong>" for i in range(len(esc))
                )
            else:
                line_html = html.escape(line)
            out.append(line_html + "<br/>")
        lines.append("<p>" + "".join(out) + "</p>")
    return "".join(lines) if lines else "<p></p>"


def create_page(
    base_url: str,
    email: str,
    api_token: str,
    space_key: str,
    title: str,
    body_markdown: str,
    parent_page_id: str | None = None,
) -> dict[str, Any]:
    base = base_url.rstrip("/")
    url = f"{base}/rest/api/content"
    storage = markdown_to_storage_html(body_markdown)
    body: dict[str, Any] = {
        "type": "page",
        "title": title,
        "space": {"key": space_key},
        "body": {
            "storage": {
                "value": storage,
                "representation": "storage",
            }
        },
    }
    if parent_page_id:
        body["ancestors"] = [{"id": parent_page_id}]
    r = requests.post(
        url,
        headers={
            "Accept": "application/json",
            "Content-Type": "application/json",
            "Authorization": _basic_auth(email, api_token),
        },
        data=json.dumps(body),
        timeout=120,
    )
    r.raise_for_status()
    return r.json()

File: web/lib/confluence.ts
import { normalizeConfluenceBaseUrl } from "./atlassian-urls";
import { abortAfter } from "./fetch-timeout";

function basicAuthHeader(email: string, token: string): string {
  const raw = Buffer.from(`${email}:${token}`, "utf8").toString("base64");
  return `Basic ${raw}`;
}

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

export function markdownToStorageHtml(md: string): string {
  const blocks = md.replace(/\r\n/g, "\n").split("\n\n");
  const parts: string[] = [];
  for (const blockRaw of blocks) {
    const block = blockRaw.trim();
    if (!block) continue;
    const lines = block.split("\n");
    const lineParts: string[] = [];
    for (const line of lines) {
      const segs = line.split("**");
      let lineHtml: string;
      if (segs.length % 2 === 1) {
        lineHtml = segs
          .map((p, i) => (i % 2 === 0 ? escapeHtml(p) : `<strong>${escapeHtml(p)}</strong>`))
          .join("");
      } else {
        lineHtml = escapeHtml(line);
      }
      lineParts.push(lineHtml + "<br/>");
    }
    parts.push("<p>" + lineParts.join("") + "</p>");
  }
  return parts.length ? parts.join("") : "<p></p>";
}

export async function confluenceHandshake(
  baseUrl: string,
  email: string,
  apiToken: string
): Promise<Record<string, unknown>> {
  const base = normalizeConfluenceBaseUrl(baseUrl);
  const r = await fetch(`${base}/rest/api/user/current`, {
    headers: {
      Accept: "application/json",
      Authorization: basicAuthHeader(email, apiToken),
    },
    signal: abortAfter(30_000),
  });
  if (!r.ok) {
    const t = await r.text();
    throw new Error(`Confluence ${r.status}: ${t.slice(0, 500)}`);
  }
  return r.json() as Promise<Record<string, unknown>>;
}

export async function createPage(
  baseUrl: string,
  email: string,
  apiToken: string,
  spaceKey: string,
  title: string,
  bodyMarkdown: string,
  parentPageId?: string | null
): Promise<Record<string, unknown>> {
  const base = normalizeConfluenceBaseUrl(baseUrl);
  const storage = markdownToStorageHtml(bodyMarkdown);
  const body: Record<string, unknown> = {
    type: "page",
    title,
    space: { key: spaceKey },
    body: {
      storage: {
        value: storage,
        representation: "storage",
      },
    },
  };
  if (parentPageId) body.ancestors = [{ id: parentPageId }];
  const r = await fetch(`${base}/rest/api/content`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      Authorization: basicAuthHeader(email, apiToken),
    },
    body: JSON.stringify(body),
    signal: abortAfter(120_000),
  });
  if (!r.ok) {
    const t = await r.text();
    throw new Error(`Confluence ${r.status}: ${t.slice(0, 500)}`);
  }
  return r.json() as Promise<Record<string, unknown>>;
}
